shtab/__init__.py

```diff
@@ -714,27 +716,33 @@ def recurse_parser(cparser, positional_idx, requirements=None):
                 if arg.choices:
                     nlist.append('( {}echo "{}" || false )'.format(
                         ' && '.join(checks + ['']),                 # Append the separator
-                        '\\n'.join(arg.choices),
+                        r'\n'.join(arg.choices),
                     ))
 
             # Ugly hack
             specials.append("'p@{}@`set cmd=($COMMAND_LINE); {}`@'".format(
                 str(idx), ' || '.join(nlist)))
 
+    if len(optionals_double) > 0:
+        if len(optionals_single) == 0:
+            # Hack: Don't add a space after completing "--" from "-"
+            optionals_single = ['-', '-']
+        else:
+            optionals_single.add('-')
     return Template("""\
 # AUTOMATICALLY GENERATED by `shtab`
 
 ${preamble}
 
 complete ${prog} \\
         'c/--/(${optionals_double_str})/' \\
-        'c/-/(${optionals_single_str} -)/' \\
+        'c/-/(${optionals_single_str})/' \\
         ${optionals_special_str} \\
         'p/*/()/'""").safe_substitute(
         preamble=("\n# Custom Preamble\n" + preamble +
                   "\n# End Custom Preamble\n" if preamble else ""), root_prefix=root_prefix,
-        prog=parser.prog, optionals_double_str=' '.join(optionals_double),
-        optionals_single_str=' '.join(optionals_single),
+        prog=parser.prog, optionals_double_str=' '.join(sorted(optionals_double)),
+        optionals_single_str=' '.join(sorted(optionals_single)),
```

Comment:
why sort these? Python's [dicts are insertion ordered](https://peps.python.org/pep-0468/), meaning you can call `.add_argument()` in the order you want things to appear.

Reply:
It was only so that the generated file was better looking. I'll remove those.